packages/wallet/api/src/marshal-contexts.js

```diff
@@ -0,0 +1,257 @@
+// @ts-check
+import { makeScalarMap } from '@agoric/store';
+import { Far, makeMarshal, Remotable } from '@endo/marshal';
+import { HandledPromise } from '@endo/eventual-send'; // TODO: convince tsc this isn't needed
+
+const { details: X, quote: q } = assert;
+
+/**
+ * Make context for exporting wallet data where brands etc. can be recognized by boardId.
+ *
+ * Objects should be registered before serialization:
+ *   - for closely-held object, use initPurseId etc.
+ *   - for published objects such as brands and issuers, use initBoardId / ensureBoardId
+ *
+ * Unregistered objects (such as instances in amounts) get serialized
+ * with a fresh slot.
+ */
+export const makeExportContext = () => {
+  const myData = {
+    purse: {
+      /** @type {MapStore<string, Purse>} */
+      bySlot: makeScalarMap(),
+      /** @type {MapStore<Purse, string>} */
+      byVal: makeScalarMap(),
+    },
+    payment: {
+      /** @type {MapStore<string, Payment>} */
+      bySlot: makeScalarMap(),
+      /** @type {MapStore<Payment, string>} */
+      byVal: makeScalarMap(),
+    },
+    // TODO: offer, contact, dapp
+  };
+  const sharedData = {
+    /** @type {MapStore<string, unknown>} */
+    bySlot: makeScalarMap(),
+    /** @type {MapStore<unknown, string>} */
+    byVal: makeScalarMap(),
+  };
+
+  /**
+   * @param {string} slot
+   * @param {string} _iface
+   */
+  const slotToVal = (slot, _iface) => {
+    if (sharedData.bySlot.has(slot)) {
+      return sharedData.bySlot.get(slot);
+    }
+    const kind = Object.keys(myData).find(k => slot.startsWith(k));
+    assert(kind, X`bad slot kind: ${slot}`);
+    const id = slot.slice(kind.length + 1);
+    const val = myData[kind].bySlot.get(id); // or throw
+    return val;
+  };
+
+  let unknownNonce = 0;
+
+  const valToSlot = val => {
+    if (sharedData.byVal.has(val)) {
+      return sharedData.byVal.get(val);
+    }
+    for (const kind of Object.keys(myData)) {
+      if (myData[kind].byVal.has(val)) {
+        const id = myData[kind].byVal.get(val);
+        return `${kind}:${id}`;
+      }
+    }
+    unknownNonce += 1;
+    const slot = `unknown:${unknownNonce}`;
+    sharedData.byVal.init(val, slot);
+    sharedData.bySlot.init(slot, val);
+    return slot;
+  };
+
+  const makeSaver = (kind, tables) => {
+    let nonce = 0;
+    return val => {
+      const slot = `${(nonce += 1)}`;
+      tables.bySlot.init(slot, val);
+      tables.byVal.init(val, slot);
```

Comment:
the `tables` param reads like something with arbitrary size. Please make it a singular of some type (per comment above)

I see below there is `table.bySlot`